File: julia/conservation.jl
#!/usr/bin/env julia
# ═══════════════════════════════════════════════════════════════
# SuperInstance Conservation Law — Julia Implementation
# γ + η = C (Shannon chain rule: H(X) = I(X;G) + H(X|G))
#
# Julia advantages: multiple dispatch, @threads parallelism,
# @inbounds/@fastmath, JIT compilation = C speed with Python syntax
# ═══════════════════════════════════════════════════════════════

module Conservation

export conservation_delta, conservation_efficiency
export fleet_cancellation, monte_carlo_cancellation
export ternary_dotproduct, ternary_matmul
export conservation_entropy, conservation_analyze
export haar_decompose, ternary_signals

const LOG2_3 = log2(3)

"""
    conservation_delta(n::Int) -> Float64

Theoretical cancellation delta: δ(n) = (1/√n)(1 - 3/(2n))
"""
function conservation_delta(n::Int)::Float64
    n < 2 && return 1.0
    return (1.0 / sqrt(n)) * (1.0 - 3.0 / (2.0 * n))
end

"""
    conservation_efficiency(n::Int) -> Float64

Conservation efficiency = 1 - δ(n)
"""
conservation_efficiency(n::Int)::Float64 = 1.0 - conservation_delta(n)

"""
    ternary_signals(n::Int) -> Vector{Int8}

Generate n random ternary signals {-1, 0, +1} with uniform probability.
"""
function ternary_signals(n::Int)::Vector{Int8}
    return Int8[rand([-1, 0, 1]) for _ in 1:n]
end

"""
    ternary_dotproduct(a::Vector{Int8}, b::Vector{Int8}) -> Int64

Branchless ternary dot product.
"""
function ternary_dotproduct(a::Vector{Int8}, b::Vector{Int8})::Int64
    n = length(a)
    s = zero(Int64)
    @inbounds for i in 1:n
        s += a[i] * b[i]
    end
    return s
end

"""
    ternary_matmul(A::Matrix{Int8}, B::Matrix{Int8}) -> Matrix{Int64}

Cache-blocked ternary matrix multiply.
"""
function ternary_matmul(A::Matrix{Int8}, B::Matrix{Int8})::Matrix{Int64}
    M, K = size(A)
    K2, N = size(B)
    @assert K == K2 "Dimension mismatch"
    C = zeros(Int64, M, N)
    
    block = 64  # Cache block size
    @inbounds for ii in 1:block:M
        iend = min(ii + block - 1, M)
        for jj in 1:block:N
            jend = min(jj + block - 1, N)
            for kk in 1:block:K
                kend = min(kk + block - 1, K)
                for i in ii:iend
                    for j in jj:jend
                        s = zero(Int64)
                        @fastmath for k in kk:kend
                            s += A[i,k] * B[k,j]
                        end
                        C[i,j] += s
                    end
                end
            end
        end
    end
    return C
end

"""
    fleet_cancellation(signals::Vector{Int8}) -> Float64

Fleet cancellation factor: 1 - |Σ signals| / n
"""
function fleet_cancellation(signals::Vector{Int8})::Float64
    n = length(signals)
    n == 0 && return 0.0
    s = zero(Int64)
    @inbounds for i in 1:n
        s += signals[i]
    end
    return 1.0 - abs(s) / n
end

"""
    conservation_entropy(signals::Vector{Int8}) -> Float64

Shannon entropy H(X) for ternary distribution {-1, 0, +1}.
Maximum: log₂(3) ≈ 1.585 bits.
"""
function conservation_entropy(signals::Vector{Int8})::Float64
    n = length(signals)
    n == 0 && return 0.0
    
    cnt_neg = cnt_zero = cnt_pos = 0
    @inbounds for i in 1:n
        if signals[i] == -1; cnt_neg += 1
        elseif signals[i] == 0; cnt_zero += 1
        else; cnt_pos += 1; end
    end
    
    H = 0.0
    for cnt in (cnt_neg, cnt_zero, cnt_pos)
        p = cnt / n
        if p > 0
            H -= p * log2(p)
        end
    end
    return H
end

"""
    conservation_analyze(X::Vector{Int8}, G::Vector{Int8}) -> NamedTuple

Conservation analysis: γ + η = C
Returns (gamma=I(X;G), eta=H(X|G), C=H(X))
"""
function conservation_analyze(X::Vector{Int8}, G::Vector{Int8})
    n = length(X)
    C_val = conservation_entropy(X)
    H_G = conservation_entropy(G)
    
    # Joint entropy H(X, G)
    joint = zeros(Int, 3, 3)
    @inbounds for i in 1:n
        joint[X[i]+2, G[i]+2] += 1
    end
    
    H_XG = 0.0
    for xi in 1:3, gi in 1:3
        p = joint[xi, gi] / n
        if p > 0
            H_XG -= p * log2(p)
        end
    end
    
    eta = max(0.0, H_XG - H_G)
    gamma = max(0.0, C_val - eta)
    
    return (gamma=gamma, eta=eta, C=C_val, H_max=LOG2_3)
end

"""
    monte_carlo_cancellation(n_agents::Int, n_trials::Int) -> Float64

Parallel Monte Carlo fleet cancellation using Threads.@threads.
"""
function monte_carlo_cancellation(n_agents::Int, n_trials::Int)::Float64
    total_cancel = zeros(Float64, Threads.nthreads())
    
    Threads.@threads for t in 1:n_trials
        tid = Threads.threadid()
        signals = Int8[rand([-1, 0, 1]) for _ in 1:n_agents]
        total_cancel[tid] += fleet_cancellation(signals)
    end
    
    return sum(total_cancel) / n_trials
end

"""
    haar_decompose(signal::Vector) -> NamedTuple

Single-level Haar wavelet decomposition.
approximation = (even + odd) / √2
detail = (even - odd) / √2
"""
function haar_decompose(signal)
    n = length(signal)
    half = n ÷ 2
    approx = Vector{Float64}(undef, half)
    detail = Vector{Float64}(undef, half)
    inv_sqrt2 = 1.0 / sqrt(2.0)
    
    @inbounds for i in 1:half
        a = Float64(signal[2i - 1])
        b = Float64(signal[2i])
        approx[i] = (a + b) * inv_sqrt2
        detail[i] = (a - b) * inv_sqrt2
    end
    
    return (approx=approx, detail=detail)
end

end # module

using .Conservation
using Printf
using Random

# ═══ Main Benchmark (run as script) ═══
if abspath(PROGRAM_FILE) == @__FILE__
    println("═══ SuperInstance Conservation Law — Julia ═══")
    println("Julia threads: $(Threads.nthreads())")
    println()
    
    println("─── Monte Carlo Fleet Cancellation ───")
    sizes = [5, 10, 50, 100, 500, 1000, 5000, 10000]
    @printf("%-8s %-12s %-12s %-8s %-10s\n", "Fleet", "Empirical", "Theory", "Error%", "Time(ms)")
    println(repeat("-", 55))
    
    for n in sizes
        t0 = time()
        mc = monte_carlo_cancellation(n, 10000)
        t1 = time()
        
        theo = conservation_efficiency(n)
        err = abs(mc - theo) / theo * 100
        
        @printf("%-8d %-12.4f %-12.4f %-8.2f %-10.1f\n",
                n, mc, theo, err, (t1-t0)*1000)
    end
    
    # Conservation identity
    println()
    println("─── Conservation Identity γ + η = C ───")
    Random.seed!(42)
    G = ternary_signals(10000)
    X = Int8[g == 1 && rand() < 0.5 ? 1 : rand([-1, 0, 1]) for (i, g) in enumerate(G)]
    tc = conservation_analyze(X, G)
    @printf("γ = %.6f bits\n", tc.gamma)
    @printf("η = %.6f bits\n", tc.eta)
    @printf("C = %.6f bits\n", tc.C)
    @printf("γ + η = %.6f ≈ C = %.6f %s\n", 
            tc.gamma + tc.eta, tc.C,
            abs(tc.gamma + tc.eta - tc.C) < 1e-10 ? "✓" : "✗")
    
    # Stress test
    println()
    println("─── Stress Test: 1,000,000 agents ───")
    t0 = time()
    mc = monte_carlo_cancellation(1_000_000, 10)
    t1 = time()
    @printf("Cancellation: %.4f%% in %.3fs\n", mc * 100, t1 - t0)
    
    println()
    println("═══ Julia Complete ═══")
end

File: fortran/conservation.f90
! ═══════════════════════════════════════════════════════════════════
! SuperInstance Conservation Law — Fortran 90 Implementation
! γ + η = C (Shannon chain rule: H(X) = I(X;G) + H(X|G))
!
! Fortran advantages: array syntax, column-major storage, OpenMP,
! pure functions, cache-friendly sequential access. The original HPC language.
! ═══════════════════════════════════════════════════════════════════

module conservation_mod
  use iso_fortran_env, only: int8, int64, real64
  implicit none
  private

  real(real64), parameter :: LOG2_3 = log(3.0_real64) / log(2.0_real64)

  public :: conservation_delta, conservation_efficiency
  public :: fleet_cancellation, monte_carlo_cancellation
  public :: ternary_dotproduct, ternary_matmul
  public :: conservation_entropy, conservation_analyze
  public :: haar_decompose
  public :: random_choice

contains

  ! δ(n) = (1/√n)(1 - 3/(2n))
  pure function conservation_delta(n) result(delta)
    integer, intent(in) :: n
    real(real64) :: delta
    if (n < 2) then
      delta = 1.0_real64
    else
      delta = (1.0_real64 / sqrt(real(n, real64))) * &
              (1.0_real64 - 3.0_real64 / (2.0_real64 * real(n, real64)))
    end if
  end function

  pure function conservation_efficiency(n) result(eff)
    integer, intent(in) :: n
    real(real64) :: eff
    eff = 1.0_real64 - conservation_delta(n)
  end function

  ! Fleet cancellation: 1 - |Σ signals| / n
  function fleet_cancellation(signals, n) result(cancel)
    integer, intent(in) :: n
    integer(int8), intent(in) :: signals(n)
    real(real64) :: cancel
    integer(int64) :: s
    integer :: i
    s = 0
    !$omp simd reduction(+:s)
    do i = 1, n
      s = s + int(signals(i), int64)
    end do
    cancel = 1.0_real64 - abs(real(s, real64)) / real(n, real64)
  end function

  ! Monte Carlo fleet cancellation
  function monte_carlo_cancellation(n_agents, n_trials) result(mean_cancel)
    integer, intent(in) :: n_agents, n_trials
    real(real64) :: mean_cancel
    real(real64) :: total_cancel
    integer :: t, i
    integer(int8), allocatable :: signals(:)
    integer :: seed_size
    integer, allocatable :: seed(:)

    call random_seed(size=seed_size)
    allocate(seed(seed_size))
    seed = 42
    call random_seed(put=seed)
    deallocate(seed)

    allocate(signals(n_agents))
    total_cancel = 0.0_real64

    !$omp parallel do reduction(+:total_cancel) private(signals, i)
    do t = 1, n_trials
      do i = 1, n_agents
        ! Random ternary: -1, 0, 1
        call random_choice(signals(i))
      end do
      total_cancel = total_cancel + fleet_cancellation(signals, n_agents)
    end do
    !$omp end parallel do

    mean_cancel = total_cancel / real(n_trials, real64)
    deallocate(signals)
  end function

  subroutine random_choice(val)
    integer(int8), intent(out) :: val
    real(real64) :: r
    call random_number(r)
    if (r < 1.0_real64 / 3.0_real64) then
      val = -1_int8
    else if (r < 2.0_real64 / 3.0_real64) then
      val = 0_int8
    else
      val = 1_int8
    end if
  end subroutine

  ! Ternary dot product
  pure function ternary_dotproduct(a, b, n) result(dp)
    integer, intent(in) :: n
    integer(int8), intent(in) :: a(n), b(n)
    integer(int64) :: dp
    integer :: i
    dp = 0
    !$omp simd reduction(+:dp)
    do i = 1, n
      dp = dp + int(a(i) * b(i), int64)
    end do
  end function

  ! Ternary matrix multiply (cache-blocked)
  subroutine ternary_matmul(A, B, C, M, K, N)
    integer, intent(in) :: M, K, N
    integer(int8), intent(in) :: A(M, K), B(K, N)
    integer(int64), intent(out) :: C(M, N)
    integer :: i, j, p
    C = 0_int64
    do i = 1, M
      do j = 1, N
        do p = 1, K
          C(i, j) = C(i, j) + int(A(i, p) * B(p, j), int64)
        end do
      end do
    end do
  end subroutine

  ! Shannon entropy for ternary
  function conservation_entropy(signals, n) result(H)
    integer, intent(in) :: n
    integer(int8), intent(in) :: signals(n)
    real(real64) :: H
    integer :: i, cnt_neg, cnt_zero, cnt_pos
    real(real64) :: p

    cnt_neg = 0; cnt_zero = 0; cnt_pos = 0
    !$omp simd reduction(+:cnt_neg, cnt_zero, cnt_pos)
    do i = 1, n
      if (signals(i) == -1_int8) cnt_neg = cnt_neg + 1
      if (signals(i) == 0_int8) cnt_zero = cnt_zero + 1
      if (signals(i) == 1_int8) cnt_pos = cnt_pos + 1
    end do

    H = 0.0_real64
    p = real(cnt_neg, real64) / real(n, real64)
    if (p > 0) H = H - p * log(p) / log(2.0_real64)
    p = real(cnt_zero, real64) / real(n, real64)
    if (p > 0) H = H - p * log(p) / log(2.0_real64)
    p = real(cnt_pos, real64) / real(n, real64)
    if (p > 0) H = H - p * log(p) / log(2.0_real64)
  end function

  ! Conservation analysis: γ + η = C
  subroutine conservation_analyze(X, G, n, gamma_val, eta_val, C_val)
    integer, intent(in) :: n
    integer(int8), intent(in) :: X(n), G(n)
    real(real64), intent(out) :: gamma_val, eta_val, C_val
    integer :: i, j
    integer :: joint(3, 3)
    real(real64) :: p, H_XG, H_G
    integer(int8), allocatable :: G_copy(:)

    C_val = conservation_entropy(X, n)
    H_G = conservation_entropy(G, n)

    ! Joint distribution
    joint = 0
    do i = 1, n
      joint(int(X(i)) + 2, int(G(i)) + 2) = &
        joint(int(X(i)) + 2, int(G(i)) + 2) + 1
    end do

    H_XG = 0.0_real64
    do i = 1, 3
      do j = 1, 3
        p = real(joint(i, j), real64) / real(n, real64)
        if (p > 0) H_XG = H_XG - p * log(p) / log(2.0_real64)
      end do
    end do

    eta_val = max(0.0_real64, H_XG - H_G)
    gamma_val = max(0.0_real64, C_val - eta_val)
  end subroutine

  ! Haar wavelet decomposition
  subroutine haar_decompose(signal, n, approx, detail)
    integer, intent(in) :: n
    integer(int8), intent(in) :: signal(n)
    real(real64), intent(out) :: approx(n/2), detail(n/2)
    integer :: i, half
    half = n / 2
    !$omp simd
    do i = 1, half
      approx(i) = (real(signal(2*i-1), real64) + real(signal(2*i), real64)) / sqrt(2.0_real64)
      detail(i) = (real(signal(2*i-1), real64) - real(signal(2*i), real64)) / sqrt(2.0_real64)
    end do
  end subroutine

end module conservation_mod

! ═══ Main Program ═══
program conservation_bench
  use iso_fortran_env, only: int8, int64, real64
  use conservation_mod
  use omp_lib
  implicit none

  integer, parameter :: DP = kind(1.0d0)
  integer :: fleet_sizes(8)
  integer :: i, n, n_trials
  real(real64) :: mc_result, theory, err, t0, t1
  real(real64) :: gamma_v, eta_v, C_v
  integer(int8), allocatable :: X(:), G(:)
  real(real64), allocatable :: approx(:), detail(:)
  integer :: seed_size
  integer, allocatable :: seed(:)

  print *, "═══ SuperInstance Conservation Law — Fortran 90 ═══"
  write(*,'(A,I0)') " OpenMP threads: ", omp_get_max_threads()
  print *, ""

  ! Monte Carlo benchmark
  print *, "─── Monte Carlo Fleet Cancellation ───"
  fleet_sizes = [5, 10, 50, 100, 1000, 10000, 100000, 1000000]
  n_trials = 10000

  write(*,'(A8, A14, A14, A10, A12)') "Fleet", "Empirical", "Theory", "Error%", "Time(ms)"

  do i = 1, size(fleet_sizes)
    n = fleet_sizes(i)
    if (n > 10000) n_trials = 100  ! reduce for large fleets
    if (n > 100000) n_trials = 10

    t0 = omp_get_wtime()
    mc_result = monte_carlo_cancellation(n, n_trials)
    t1 = omp_get_wtime()

    theory = conservation_efficiency(n)
    err = abs(mc_result - theory) / theory * 100.0_real64

    write(*,'(I8, F14.4, F14.4, F10.2, F12.1)') &
      n, mc_result, theory, err, (t1 - t0) * 1000.0_real64
  end do

  ! Conservation identity
  print *, ""
  print *, "─── Conservation Identity γ + η = C ───"
  n = 10000
  allocate(X(n), G(n))
  call random_seed(size=seed_size)
  allocate(seed(seed_size))
  seed = 123
  call random_seed(put=seed)
  deallocate(seed)

  do i = 1, n
    call random_choice(X(i))
    call random_choice(G(i))
  end do

  call conservation_analyze(X, G, n, gamma_v, eta_v, C_v)
  write(*,'(A, F10.6)') " γ = ", gamma_v
  write(*,'(A, F10.6)') " η = ", eta_v
  write(*,'(A, F10.6)') " C = ", C_v
  write(*,'(A, F10.6)') " γ+η = ", gamma_v + eta_v

  ! Haar wavelet
  print *, ""
  print *, "─── Haar Wavelet Decomposition ───"
  deallocate(X)
  allocate(X(8))
  X = [1_int8, 1_int8, -1_int8, 1_int8, -1_int8, -1_int8, 1_int8, -1_int8]
  allocate(approx(4), detail(4))
  call haar_decompose(X, 8, approx, detail)
  write(*,'(A, 4F8.3)') " Approx: ", approx
  write(*,'(A, 4F8.3)') " Detail: ", detail

  deallocate(X, G, approx, detail)
  print *, ""
  print *, "═══ Fortran Complete ═══"

end program conservation_bench
